### ISSUE TITLE
About training stability

### ISSUE THREAD

**Issue by wonderseen** [Dao-AILab/flash-attention#686]
Hi team, 

With the flash_attn_unpadded_func implementation [flash_atten==0.2.8], we ran the same training trial twice to check its training stability.

- At training:
we found that these two training loss curves were not exactly the same during the training process.

- After training:
Though these two trained model also achieved not-bad performances on benchmarks, there are some difference between their performances with respect to various testsets.

I don't know whether it is reasonable for flash_atten? Or any suggestions to avoid training instability when using flash_atten?

Thanks.

**Comment by tridao** [comment]
When you say "stability", do you mean the loss explodes / NaNs for one implementation and not another?

Is the difference between the two implementation be larger / smaller / similar to changing the random seed?

You can use whichever implementation that works better for you.

**Comment by wonderseen** [comment]
> When you say "stability", do you mean the loss explodes / NaNs for one implementation and not the other?
> 
> Is the difference between the two implementation larger / smaller / similar to changing the random seed?
> 
> You can use whichever implementation that works better for you.

thanks for the quick reply. 

![test](https://github.com/Dao-AILab/flash-attention/assets/31044791/d9654164-36f6-4d83-9cfa-243cbee3f583)


"stability": I means that two models were both trained well using exactly the same hyperparameters (the random seeds were also the same), but their losses and performance are not the same at the same training step (with some slightly difference as shown in this figure).

At this case, if we further retrain a new model using a new set of data and the new model show marginal gains on benchmarks, it is hard to determine whether the benefits come from **training instability** or **data optimization**.

**Comment by tridao** [comment]
I see, I usually use stability to refer to whether things are stable (e.g. loss explosion).
FlashAttention gives the same output and gradient as standard implementation, up to numerical precision. It's been used to train large-scale models (e.g. Pythia, Llama). Pythia folks saw the same training curves between different implementations.

If you're still conservative, you don't have to use it and you can stick with the standard implementation.

**Comment by wonderseen** [comment]
> I see, I usually use stability to refer to whether things are stable (e.g. loss explosion). FlashAttention gives the same output and gradient as standard implementation, up to numerical precision. It's been used to train large-scale models (e.g. Pythia, Llama). Pythia folks saw the same training curves between different implementations.
> 
> If you're still conservative, you don't have to use it and you can stick with the standard implementation.

There may be some misunderstanding here.

The figure above records the loss training curves of two runs both using flash attention with the same setting.  (flash attention run-1 vs flash attention run-2)

We faced similar situations before using standard implementation to train llama [trained a model twice, got different model accuracy on generation tasks]. Some warning told that  `TF_ENABLE_ONEDNN_OPTS=0` might help.
``` 
oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
```

I am not sure whether the order of computation causes slightly mismatched results between different runs using flash attention. Do we have any variable or suggestions to control the order of calculation? Thanks~

**Comment by tridao** [comment]
Oh I see. The backward pass of FlashAttention is not deterministic due to atomic adds. We do plan to eventually add an option for deterministic backward (at the cost of running slower) but I haven't had the bandwidth to work on that.

**Comment by wonderseen** [comment]
> Oh I see. The backward pass of FlashAttention is not deterministic due to atomic adds. We do plan to eventually add an option for deterministic backward (at the cost of running slower) but I haven't had the bandwidth to work on that.

Appreciate it! Looking forward to the updates.
